lib/src/java/facter.cc

```diff
@@ -0,0 +1,155 @@
+#include "com_puppetlabs_Facter.h"
+#include <facter/facts/collection.hpp>
+#include <facter/facts/scalar_value.hpp>
+#include <facter/facts/array_value.hpp>
+#include <facter/facts/map_value.hpp>
+#include <string>
+
+using namespace std;
+using namespace facter::facts;
+
+static jclass object_class, long_class, double_class, boolean_class, hash_class;
+static jmethodID long_constructor, double_constructor, boolean_constructor, hash_constructor, hash_put;
+
+static string to_string(JNIEnv* env, jstring str)
+{
+    if (!str) {
+        return {};
+    }
+
+    auto ptr = env->GetStringUTFChars(str, nullptr);
+    if (!ptr) {
+        return {};
+    }
+
+    auto size = env->GetStringUTFLength(str);
+    string result(ptr, size);
+
+    env->ReleaseStringUTFChars(str, ptr);
+    return result;
+}
+
+static jobject to_object(JNIEnv* env, value const* val)
+{
+    if (!val) {
+        return nullptr;
+    }
+    if (auto ptr = dynamic_cast<string_value const*>(val)) {
```

Comment:
Yeah (assuming it was a comment on the `dynamic_cast`), in retrospect I would have used a `boost::variant` rather than a class hierarchy.  The thought originally was to allow third parties to provide their own "values", whatever they are.  I think in the end we'll just support primitives, arrays, and hashes anyway, so having an abstract base class is entirely unnecessary.